packages/components/src/index.ts

```diff
@@ -12,7 +12,10 @@ export {
 } from '@wordpress/primitives';
 
 // Components.
-export { default as __experimentalAlignmentMatrixControl } from './alignment-matrix-control';
+export {
+	default as __experimentalAlignmentMatrixControl,
```

Comment:
While I appreciate that we're keeping the `__experimental` exports for backward compatibility reasons, I wonder if we should deprecate `__experimental` imports now that the components are stable. That will nudge 3rd party developers to migrate to stable imports, otherwise they might be stuck with the `__experimental` ones forever

Reply:
There are so many "shades" of deprecation — hard, soft, TS-only, TSDoc only... 😂 For these I think a TSDoc <code>@deprecated Import `AlignmentMatrixControl` instead</code> would be appropriate. Is that what you had in mind?

Reply:
Yup, it's odd that deprecation can be so complex 😅 

Anyway, imagine I'm a plugin dev that uses some experimental components in my plugin. How will a TSDoc `@deprecated` help me find out that I'm using a deprecated import and I have to replace it with the canonical stable import? Will I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

Reply:
> ill I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

I managed to make the `@deprecate` annotation work when it's added just before the actual export symobl, inside the `export` block:

```typescript
export {
	/**
	 * @deprecated Since version X.X. Will be removed in version X.Y. Use `ConfirmDialog` instead.
	 */
	ConfirmDialog as __experimentalConfirmDialog,
	ConfirmDialog,
} from './confirm-dialog';
```

I can confirm it doesn't affect the second export there:

<img width="773" alt="Screenshot 2024-04-18 at 12 51 30 p m" src="https://github.com/WordPress/gutenberg/assets/81248/021372f5-f740-4e75-93ab-8dc63b7ac704">

And of the `@deprecated` declartion, it will render like this in `vscode`:

| | |
|-|-|
| <img width="974" alt="Screenshot 2024-04-18 at 12 48 28 p m" src="https://github.com/WordPress/gutenberg/assets/81248/4ca193e0-4fda-446a-b2d3-7ec8f61e0412"> | <img width="944" alt="Screenshot 2024-04-18 at 12 49 04 p m" src="https://github.com/WordPress/gutenberg/assets/81248/03a20d2b-78ee-4782-837d-b7a246795ce8"> |

So I assume all editors/IDEs that support TSDoc will warn users about it.

If that's a good way to prevent developers from continuing to use it, that is another story. It might be good to do the deprecation in gradually? First soft-deprecate using a tsdoc, announce in the changelog and perhaps a Make post, then in another (major?) version, completely remove the `__experimental` exports.

Reply:
> Anyway, imagine I'm a plugin dev that uses some experimental components in my plugin. How will a TSDoc `@deprecated` help me find out that I'm using a deprecated import and I have to replace it with the canonical stable import? Will I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

My current stance on this is, 

1. Do we plan to hard deprecate it?
2. Does switching to the new thing provide the consumer with actual benefits over the old thing?

If the answer is "no" to both questions, we shouldn't log a console warning and we shouldn't block the TS from compiling. We shouldn't demand a dev's attention for what is essentially a housekeeping detail. If they care about housekeeping details, they'll do it.

And to be clear, personally I don't think the `__experimental` exports are worth hard deprecating. It's virtually free to maintain this BC, compared to the cost on the ecosystem if we yank them.

Reply:
Oh, the complex deprecation discussion again 😅 

I think we have far too many cases that are cheap/free to maintain, but because we have too many, it adds so much noise. I wish we had a clear deprecation policy that actually allows us to get rid of deprecated features at some point. 

If I come back to planet Earth though and draw a line on the tradeoffs, you're right and it should be just fine to continue soft-deprecating as suggested.